src/viadot/sources/sftp.py

```diff
@@ -0,0 +1,296 @@
+
+        Returns:
+            pd.DataFrame: The response data as a Pandas Data Frame plus viadot metadata.
+        """
+        byte_file = self._getfo_file(file_name=file_name)
+        byte_file.seek(0)
+
+        self._close_conn()
+
+        if Path(file_name).suffix == ".csv":
+            df = pd.read_csv(byte_file, sep=sep, usecols=columns)
+
+        elif Path(file_name).suffix == ".parquet":
+            df = pd.read_parquet(byte_file, usecols=columns)
+
+        elif Path(file_name).suffix == ".tsv":
+            df = pd.read_csv(byte_file, sep=sep, usecols=columns)
+
+        elif Path(file_name).suffix in [".xls", ".xlsx", ".xlsm"]:
+            df = pd.read_excel(byte_file, usecols=columns)
+
+        elif Path(file_name).suffix == ".json":
+            df = pd.read_json(byte_file)
+
+        elif Path(file_name).suffix == ".pkl":
+            df = pd.read_pickle(byte_file)
+
+        elif Path(file_name).suffix == ".sql":
+            df = pd.read_sql(byte_file)
+
+        elif Path(file_name).suffix == ".hdf":
+            df = pd.read_hdf(byte_file)
+
+        else:
+            message = (
+                f"Not able to read the file {Path(file_name).name}, "
+                + f"unsupported filetype: {Path(file_name).suffix}"
+            )
+            raise ValueError(message)
+
+        if df.empty:
+            self._handle_if_empty(
+                if_empty=if_empty,
+                message="The response does not contain any data.",
+            )
+        else:
+            self.logger.info("Successfully downloaded data from the SFTP server.")
+
+        return df
+
+    def _list_directory(self, path: str | None = None) -> list[str]:
+        """Returns a list of files on the remote system.
+
+        Args:
+            path (str, optional): full path to the remote directory to list.
+                Defaults to None.
+
+        Returns:
+            List: List of files.
+        """
+        path = "." if path is None else path
+
+        return self.conn.listdir(path)
+
+    def _recursive_listdir(
+        self, path: str = ".", files: defaultdict(list) = None
+    ) -> defaultdict(list):
+        """Recursively returns a defaultdict of files on the remote system.
+
+        Args:
+            path (str, optional): full path to the remote directory to list.
+                Defaults to None.
+            files (defaultdict(list), optional): parameter to call recursively.
+
+        Returns:
+            defaultdict(list): List of files.
+        """
+        if files is None:
+            files = defaultdict(list)
+
+        for attr in self.conn.listdir_attr(str(path)):
+            if S_ISDIR(attr.st_mode):
+                self._recursive_listdir(Path(path) / attr.filename, files)
+
+            else:
+                files[path].append(attr.filename)
+
+        return files
+
+    def _process_defaultdict(self, defaultdict: defaultdict(list)) -> list[str]:
+        """Process defaultdict to list of files.
+
```

Comment:
I think the name is so generic it doesn't really mean anything, would make it more specific

```suggestion
    def _files_defaultdict_to_list(self, defaultdict: defaultdict(list)) -> list[str]:
        """Convert a list of files from a defaultdict to a list.

```